tracer/src/Datadog.Trace/Debugger/LiveDebugger.cs

```diff
@@ -316,12 +317,31 @@ private void CheckUnboundProbes()
                     return;
                 }
 
+                // Initialize these lists only when there is at least one unbound probe that becomes bound, to reduce unnecessary allocations.
+                List<NativeLineProbeDefinition> lineProbes = null;
+                List<ProbeDefinition> noLongerUnboundProbes = null;
+
                 foreach (var unboundProbe in _unboundProbes)
                 {
-                    var result = _lineProbeResolver.TryResolveLineProbe(unboundProbe, out var bytecodeLocation);
+                    var result = _lineProbeResolver.TryResolveLineProbe(unboundProbe, out var location);
                     if (result.Status == LiveProbeResolveStatus.Bound)
                     {
-                        // TODO: Install the line probe.
+                        lineProbes ??= new List<NativeLineProbeDefinition>();
+                        noLongerUnboundProbes ??= new List<ProbeDefinition>();
+
+                        noLongerUnboundProbes.Add(unboundProbe);
+                        lineProbes.Add(new NativeLineProbeDefinition(location.ProbeDefinition.Id, location.MVID, location.MethodToken, (int)location.BytecodeOffset, location.LineNumber, location.ProbeDefinition.Where.SourceFile));
+                    }
+                }
+
+                if (lineProbes?.Any() == true)
+                {
+                    Log.Information<int>("LiveDebugger.CheckUnboundProbes: {Count} unbound probes became bound.", noLongerUnboundProbes.Count);
+                    DebuggerNativeMethods.InstrumentProbes(Array.Empty<NativeMethodProbeDefinition>(), lineProbes.ToArray(), Array.Empty<NativeSpanProbeDefinition>(), Array.Empty<NativeRemoveProbeRequest>());
+
+                    foreach (var boundProbe in noLongerUnboundProbes)
```

Comment:
Maybe we can clone the `unboundProbes` collection and work on a snapshot instead on the real collection? then we can remove the "no longer unboud probe" in the same place where we add it to the `lineProbes` collection.

Reply:
Considering the trade-off between cloning the entire collection and accumulating the bounded probes, I'm in favor of accumulating  because in most cases only some of probes will turn bound and most of the time the module will not be even relevant for any of the unbound probes at all so creating eagerly a copy of a collection will be somewhat inefficient.
I prefer spending some more CPU cycles and spare the memory allocation.
WDYT?